src/com/google/javascript/jscomp/parsing/parser/util/UnicodeMatch.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2020 The Closure Compiler Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.javascript.jscomp.parsing.parser.util;
+
+import java.util.regex.Pattern;
+
+/**
+ * Exposes non-ASCII unicode detection for Javascript identifiers in compliance
+ * with ECMAScript spec. Retains old optimizations for commonly-used characters,
+ * namely ASCII and Latin/Greek charsets. Uses regex for remaining characters
+ * (which would throw errors otherwise).
+ * 
+ * Unicode category patterns with /u flag compiled to vanilla regex with
+ * regexpu: https://github.com/mathiasbynens/regexpu
+ * 
+ * IdentifierStart and IdentifierPart regex patterns based on ECMA's
+ * documentation:
+ * https://github.com/tc39/proposal-regexp-unicode-property-escapes#other-examples
+ * 
+ * const regexIdentifierStart = /[$_\p{ID_Start}]/u;
+ * const regexIdentifierPart = /[$_\u200C\u200D\p{ID_Continue}]/u;
+ */
+public class UnicodeMatch {
+  // Intentional to minimize branches in this code
+  @SuppressWarnings("ShortCircuitBoolean")
+  private static boolean fastPathLetters(char ch) {
```

Comment:
Could you also copy over the comment about minimizing branch prediction misses? It's not really clear to me otherwise why we are trying to minimize branches.

About adding the Latin and Greek letters to this fast path: I'm concerned that by adding those letters to this function, we're making the more common case of ASCII characters slower due to the lack of short-circuiting.

It seems potentially useful to have a fast path for these character, if they're common in Angular, but not at the expense of ASCII characters. @concavelenz was the one who initially added this fast path - John, do you have any thoughts?

Reply:
We don't need to include non-ascii (>127) in the fast path.

Reply:
So, to be clear, I should remove the non-ASCII part of the fast path (fast path effectively becomes just `0 < ch < 127`)? 

The Latin and Greek optimizations were there before I touched the code - will adjust, but just looking for acknowledgement that I'm not misunderstanding the direction, which may impact Angular build times. I did not insert those optimizations.

Reply:
I'm admittedly confused about the note here, my apologies - it seems to me like there are a lot of ASCII chars which are not valid JS identifier characters (start or part), and the Greek/Latin fast paths are there to catch those before falling back to more expensive checks (before, I believe it was `Character.isLetter()`, now it would be regex).

Please advise, and thank you both for your patience.